File: 3.math-stats/all-combinations/combine_sum.rb
def combine_sum(candidates, target)
  res = []
  dfs(candidates, target, [], res)
  return res
end

def dfs(nums, target, path, res)
  if target == 0
    res.push(path)
    return res
  end
  for i in 0..(nums.length - 1)
    if nums[i] > target
      next
    end
    dfs(nums[i..-1], target - nums[i], path + [nums[i]], res)
  end
end

arr = [2, 3, 6, 7]
target = 7
p(combine_sum(arr, 7))

File: 3.math-stats/all-combinations/combineSum.js
const combineSum = function (candidates, target) {
  const res = [];
  dfs(candidates, target, [], res);
  return res;
};
const dfs = function (nums, target, path, res) {
  if (target === 0) {
    res.push(path);
    return res;
  }
  for (let i = 0; i < nums.length; ++i) {
    const newNums = [...nums];
    if (i > 0) {
      newNums.splice(0, i);
    }
    if (nums[i] > target) {
      continue;
    }
    dfs(newNums, target - nums[i], [...path, ...[nums[i]]], res);
  }
};

const arr = [2, 3, 6, 7];
const target = 7;
console.log(combineSum(arr, 7));
